Simplify `bin/report` implementation

Malax opened the pull request.

dzuelke commented on lib/metrics.sh:
```diff
@@ -0,0 +1,112 @@
+# Sets a metric value as raw JSON data.
+# The value parameter should be valid JSON (number, boolean, string, etc.).
+#
+# NOTE: Strings must include quotes (use `metrics::set_string` for automatic quoting).
```
```suggestion
# NOTE: Strings must be wrapped in double quotes (use `metrics::set_string` for automatic wrapping).
```

dzuelke commented on lib/metrics.sh:
```diff
@@ -0,0 +1,112 @@
+}
+
+# Sets a metric value as a string.
+# The value will be automatically quoted and escaped for JSON.
```
```suggestion
# The value will be automatically wrapped in double quotes for the final output.
```

dzuelke commented on lib/metrics.sh:
```diff
@@ -0,0 +1,112 @@
+	local key="${1}"
+	local value="${2}"
+
+	metrics::set_raw "${key}" "\"${value}\""
```
The issue with this is that it'll still break if the string contains double quotes, but you can use JQ to produce valid JSON as input for `set_raw` :)
```suggestion
	metrics::set_raw "$key" "$(jq -n --arg value "$value" '$value')"
```

Malax replied:
I really like this - proper escaping at the cost of one sub shell. Added in 2943742.

dzuelke commented on lib/metrics.sh:
````diff
@@ -0,0 +1,112 @@
+# # ... some operation ...
+# metrics::set_time "compile_duration" "${start_time}"
+# ```
+metrics::set_time() {
````
Should this not be named `set_duration()`?

Malax replied:
Yeah, good call. This is the old name of the function for the legacy metrics implementation. I didn't question the name at all, but it should be `set_duration`. Fixed via 940b562.

Malax commented:
@dzuelke I addressed the comments. I decided to keep the JSON terminology in the comments as:
1. This is an internal library
2. The abstraction would be leaky anyways
3. Saying it works with JSON values allows us to not re-state what is valid and what is not
